[FLINK-6043] [web] Display exception timestamp

zentol opened the pull request:
This PR adds the a timestamp to the exception info in the web frontend.

Here's how it looks now:
![after](https://cloud.githubusercontent.com/assets/5725237/24140505/991929c6-0e20-11e7-8254-4825b41cf8b7.png)
(before there simply was no timestamp)

The root exception timestamp is the FAILED timestamp of the first execution that failed.
For other exceptions the time when the corresponding execution switched into the FAILED state.

The root exception timestamp is stored in a separate field in the ```ExecutionGraph``` that is updated within ```fail(Throwable, long)```. I purposefully didn't use any EG timestamp (the most appropriate would be FAILING) since the displayed timestamps may be different.

We don't take the timestamp on the TM because *all* timestamps in the webUI are taken on the JM, so it could lead to odd situations where the execution failed before the execution happened.

uce commented on flink-runtime-web/web-dashboard/app/partials/jobs/job.exceptions.jade:
```diff
@@ -20,6 +20,10 @@
 
+  .panel-heading.clearfix
+    .panel-info.thin.last
+      | Timestamp: {{ exceptions['timestamp'] | amDateFormat:'YYYY-MM-DD, H:mm:ss' }}
```
I would do it like this:
```
strong Timestamp:
= ' '
| {{ exceptions['timestamp'] | amDateFormat:'YYYY-MM-DD, H:mm:ss' }}
```

![screen shot 2017-03-21 at 11 20 02](https://cloud.githubusercontent.com/assets/1756620/24142965/777d370a-0e28-11e7-86e1-d4a61c6656cc.png)

uce commented on flink-runtime-web/web-dashboard/app/partials/jobs/job.exceptions.jade:
```diff
@@ -31,7 +35,9 @@
-      span {{ exception.location }}
+      | Timestamp: {{ exceptions['timestamp'] | amDateFormat:'YYYY-MM-DD, H:mm:ss' }}
+      br
+      | Location: {{ exception.location }}
```
I would do multiple panels like this:
```
.panel-info.thin
  strong Timestamp:
  = ' '
  | {{ exceptions['timestamp'] | amDateFormat:'YYYY-MM-DD, H:mm:ss' }}
.panel-info.thin.last
  strong Location:
  = ' '
  | {{ exception.location }}
```

![screen shot 2017-03-21 at 11 20 06](https://cloud.githubusercontent.com/assets/1756620/24143019/a6105548-0e28-11e7-9d35-5522c7626845.png)

uce commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/AccessExecutionGraph.java:
```diff
@@ -70,6 +70,8 @@
 	 */
 	String getFailureCauseAsString();
 
+	long getFailureTimeStamp();
```
Missing JavaDocs. This would be the only method in this class without one...

uce replied:
`TimeStamp` with upper case `S` is inconsistent with other uses and I think incorrect. Could you replace it with `timestamp` here in other classes?

uce commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ArchivedExecutionGraph.java:
```diff
@@ -73,6 +73,8 @@
 	 */
 	private final String failureCause;
 
+	private long failureTimeStamp;
```
- Missing comment
- `failureTimeStamp` -> `failureTimestamp`

uce commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ArchivedExecutionGraph.java:
```diff
@@ -94,6 +96,7 @@ public ArchivedExecutionGraph(
 			long[] stateTimestamps,
 			JobStatus state,
 			String failureCause,
+			long failureTimeStamp,
```
And in other places as well

uce commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java:
```diff
@@ -212,6 +212,8 @@
 	 * that was not recoverable and triggered job failure */
 	private volatile Throwable failureCause;
 
+	private volatile long failureTimeStamp;
```
Missing JavaDocs

zentol commented:
@uce I've addressed your comments. The timestamp box will now be hidden if the timestamp wasn't visible yet. For the exception stuff in the EG I've introduce a separate container class ```ErrorInfo``` that holds the Exception and timestamp, which enforces that both have to be supplied.

uce commented:
Thanks for addressing the comments. I'm unsure about the `ErrorInfo` instead of adding the suggested method. Since @tillrohrmann and @StephanEwen often work on the `ExecutionGraph` I would like their input on this.

zentol commented:
The point of the ErrorInfo is to actually enforce that the exception is always set with a timestamp. By adding a method you enforce nothing, the exception can still be set without a timesmtap.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java:
```diff
@@ -1038,7 +1038,8 @@ private boolean transitionState(ExecutionState currentState, ExecutionState targ
 		if (STATE_UPDATER.compareAndSet(this, currentState, targetState)) {
-			markTimestamp(targetState);
+			long timestamp = System.currentTimeMillis();
+			markTimestamp(targetState, timestamp);
```
Can we let `markTimestamp` return the timestamp and move `System.currentTimeMillis` back to it?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ErrorInfo.java:
```diff
@@ -0,0 +1,55 @@
+ */
+class ErrorInfo {
+	private volatile Throwable exception;
+	private volatile long timestamp;
```
I'd like to make them final.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java:
```diff
@@ -210,7 +210,7 @@
 	/** The exception that caused the job to fail. This is set to the first root exception
 	 * that was not recoverable and triggered job failure */
-	private volatile Throwable failureCause;
+	private final ErrorInfo failureCause = new ErrorInfo();
```
making the fields of the `ErrorInfo` final and `failureCause` volatile is better imho.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ErrorInfo.java:
```diff
@@ -0,0 +1,55 @@
+	 */
+	void setExceptionAndTimestamp(Throwable exception, long timestamp) {
+		this.exception = exception;
+		this.timestamp = timestamp;
```
What if we don't allow the user to specify a timestamp but automatically call `System.currentTimeMillis`? That way we wouldn't have to wire the `causeTimestamp` through all the different functions in the `ExecutionGraph`.

zentol replied:
That would make the root exception inconsistent with the task exception in regards to the time.

zentol replied:
What about the following: Replace the Throwable that we write through the methods with en ErrorInfo instead.

zentol replied:
We can also add an additional constructor that only takes a throwable and creates a timestamp.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ErrorInfo.java:
```diff
@@ -0,0 +1,55 @@
+	 * @param timestamp Timestamp when exception occurred
+	 */
+	void setExceptionAndTimestamp(Throwable exception, long timestamp) {
+		this.exception = exception;
```
Doesn't this fall under the the principle of "tooling shouldn't throw exceptions?"

StephanEwen commented:
Can we remove the `errorTimestamp` from the `notifyStateTransition`? I think it obfuscates the meaning a bit...

zentol commented:
@tillrohrmann I've addressed your comments regarding ErrorInfo being volatile/fields being final as well as markTimestamp generating and returning the timestamp.

@StephanEwen @tillrohrmann Regarding the timestamp in notifyStateTransiton etc. :The webUI displays 2 kind of exceptions: 1) the root exception that is stored in the ExecutionGraph and 2) a list of task exceptions that are stored in the Executions. IMO, if the exception stored in an Execution is also the root exception in the EG then the timestamps should match to be consistent and not give off the impression that these are unrelated/different. To that end the Execution has to supply the Exception as well as the timestamp. We could create the ErrorInfo directly in the Execution and pass that on instead of just the exception. This would make it move obvious that the TS is related to the exception i suppose.

tillrohrmann commented:
Wouldn't it be ok, if the timestamps differed slightly? I mean the EG root cause timestamp is the time when the EG was failed, whereas the `Execution` related timestamp says when the failure at the `Execution` was detected, for example.

zentol commented:
It's inconsistent, that's all I'm saying.

Passing an ErrorInfo instead of a Throwable is also useful (probably even required) for FLINK-6042 since it allows us to attach meta data to the exception (task name, task location, attempt id). The attempt id is especially important so we can extract the correct exception history from the executions.

zentol commented:
Of course it wouldn't be the end of the world to display an inconsistent timestamp, it just doesn't make sense for them to be different though. The root exception display is only meant to highlight the first exception that caused the job to fail; so why should the timestamps be different.

Basically, i wanna know why it is so horrible to pass on the timestamp.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ErrorInfo.java:
```diff
@@ -0,0 +1,49 @@
+	private final long timestamp;
+	
+	ErrorInfo(Throwable exception, long timestamp) {
+		this.exception = exception;
```
`checkNotNull` is missing.

zentol replied:
Doesn't this fall under the the principle of "tooling shouldn't throw exceptions?"

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java:
```diff
@@ -632,8 +632,8 @@ void executionFailed(Throwable t) {
 	 */
-	void notifyStateTransition(ExecutionAttemptID executionId, ExecutionState newState, Throwable error) {
-		getExecutionGraph().notifyExecutionChange(getJobvertexId(), subTaskIndex, executionId, newState, error);
+	void notifyStateTransition(ExecutionAttemptID executionId, ExecutionState newState, Throwable error, long errorTimestamp) {
```
It's not strictly speaking the error timestamp, right? It's more of a state transition timestamp.

zentol replied:
Technically, yes, it is the TS when the Execution went into the state FAILED.

Since we can't use the TM timestamp this seemed like the most appropriate one to take. The only other option would be to take a TS when the UpdateTaskExecutionState message hits the JM; at that point though we are already at a point where the timestamp isn't particularly accurate, so by using the FAILED timestamp we aren't really losing accuracy, but affect fewer methods.

tillrohrmann commented:
It's not per se horrible to pass on the timestamp. In my opinion it's just a good idea trying to keep the interface as lean as possible to avoid making mistakes. If we say that we want to display a consistent timestamp, then I think we have to pass it to the `ExecutionGraph` as you've done it.

zentol commented:
@tillrohrmann What do you think about passing an ErrorInfo instead of a Throwable so that we don't need the additional timestamp argument?

zentol commented:
@tillrohrmann I've updated the PR. The changes to the EG we're reduced to a minimum; `ExecutionGraph#fail()/suspend()` now take an `ErrorInfo` instead of a `Throwable`. The old methods were kept to not require further changes in the EG.

In the EG `String getFailureCauseAsString()` was replaced with `ErrorInfo getFailureCause()`. This is basically future-proofing; for FLINK-6042 i would like to store additional meta data (task name/location) in the error info but not add methods to the EG for each of them. This requires ErrorInfo to implement Serializable.

greghogan commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ErrorInfo.java:
```diff
@@ -0,0 +1,84 @@
+	/**
+	 * Returns the contained exception.
+	 *
+	 * @return contained exception, or null if no exception was set yet
```
Perhaps reword "if no exception was set yet" to reference that the field is not deserialized?

zentol replied:
good idea.

greghogan commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java:
```diff
@@ -955,14 +950,33 @@ public void stop() throws StoppingException {
+	/**
+	 * Suspends the current ExecutionGraph.
+	 *
+	 * The JobStatus will be directly set to SUSPENDED iff the current state is not a terminal
```
Is it better to spell out "iff"?

zentol replied:
it think it's ok.

zentol commented:
I've rebased the branch, @tillrohrmann could you check it out again for the changes in the ExecutionGraph?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java:
```diff
@@ -1070,9 +1088,9 @@ public void suspend(Throwable suspensionCause) {
+	 * @param errorInfo ErrorInfo containing the exception that caused the failure.
 	 */
-	public void failGlobal(Throwable t) {
+	public void failGlobal(ErrorInfo errorInfo) {
```
Why don't we do a null check for `errorInfo` similar to the one in `suspend`? I think it should be fine to assume that this parameter is not null.

zentol commented:
@tillrohrmann Thanks for the review, will add the null check to `failGlobal` and merge it.
